File: backend/spring-boot-kotlin/src/main/java/org/bootstrapbugz/api/shared/validator/impl/FieldMatchImpl.java
package org.bootstrapbugz.api.shared.validator.impl;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import org.bootstrapbugz.api.shared.validator.FieldMatch;
import org.springframework.beans.BeanWrapperImpl;

public class FieldMatchImpl implements ConstraintValidator<FieldMatch, Object> {
  private String firstFieldName;
  private String secondFieldName;

  @Override
  public void initialize(FieldMatch constraint) {
    firstFieldName = constraint.first();
    secondFieldName = constraint.second();
  }

  @Override
  public boolean isValid(Object obj, ConstraintValidatorContext context) {
    final var wrapper = new BeanWrapperImpl(obj);
    final var first = (String) wrapper.getPropertyValue(firstFieldName);
    final var second = (String) wrapper.getPropertyValue(secondFieldName);

    return first == null && second == null || first != null && first.equals(second);
  }
}

File: backend/spring-boot-kotlin/src/main/kotlin/org/bootstrapbugz/api/shared/validator/impl/FieldMatchImpl.kt
package org.bootstrapbugz.api.shared.validator.impl

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import org.bootstrapbugz.api.shared.validator.FieldMatch
import org.springframework.beans.BeanWrapperImpl

class FieldMatchImpl : ConstraintValidator<FieldMatch, Any?> {
  private var firstFieldName: String = ""
  private var secondFieldName: String = ""

  override fun initialize(constraint: FieldMatch) {
    firstFieldName = constraint.first
    secondFieldName = constraint.second
  }

  override fun isValid(obj: Any?, context: ConstraintValidatorContext): Boolean {
    val wrapper = BeanWrapperImpl(obj)
    val first = wrapper.getPropertyValue(firstFieldName) as? String
    val second = wrapper.getPropertyValue(secondFieldName) as? String
    return first == null && second == null || first != null && first == second
  }
}
